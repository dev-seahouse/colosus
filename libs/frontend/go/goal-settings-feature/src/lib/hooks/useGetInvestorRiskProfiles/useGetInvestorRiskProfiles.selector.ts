import type { UseQueryOptions } from '@tanstack/react-query';
import useGetInvestorRiskProfiles, {
  getInvestorRiskProfilesQuery,
} from './useGetInvestorRiskProfiles';
import type {
  ConnectInvestorGetInvestorRiskProfilesResponseDto,
  ConnectInvestorRiskProfileTypes,
} from '@bambu/api-client';
import { transformRiskProfileDescription } from '../../utils/transformRiskProfileDescription/transformRiskProfileDescription';

export type InvestorRiskProfileNamesMap = {
  [key in ConnectInvestorRiskProfileTypes]: {
    name: key;
    description: string[];
  };
};

// returns investor risk profiles keyed by name, e.g data['Very Conservative']
export function useSelectInvestorRiskProfilesNamesMap(
  options?: UseQueryOptions<
    ConnectInvestorGetInvestorRiskProfilesResponseDto,
    unknown,
    InvestorRiskProfileNamesMap,
    string[]
  >
) {
  return useGetInvestorRiskProfiles({
    queryKey: getInvestorRiskProfilesQuery.queryKey,
    queryFn: getInvestorRiskProfilesQuery.queryFn,
    select: (data) => transformRiskProfiles(data),
    ...options,
  });
}

export function transformRiskProfiles(
  data: ConnectInvestorGetInvestorRiskProfilesResponseDto
): InvestorRiskProfileNamesMap {
  return data.reduce((acc: any, item) => {
    acc[item.riskProfileName] = {
      name: item.riskProfileName,
      description: transformRiskProfileDescription(item.riskProfileDescription),
    };
    return acc;
  }, {} as InvestorRiskProfileNamesMap);
}

export function transFormRiskProfileToIdMap(
  data: ConnectInvestorGetInvestorRiskProfilesResponseDto
): TransformToIdMap<ConnectInvestorGetInvestorRiskProfilesResponseDto> {
  return data.reduce((acc: any, item) => {
    acc[item.id] = {
      ...item,
    };
    return acc;
  }, {});
}

type TransformToIdMap<T extends Array<{ id: string }>> = {
  [K in T[number]['id']]: Extract<T[number], { id: K }>;
};
