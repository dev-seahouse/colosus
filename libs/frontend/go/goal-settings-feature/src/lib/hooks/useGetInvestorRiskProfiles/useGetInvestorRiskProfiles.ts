import {
  type ConnectInvestorGetInvestorRiskProfilesResponseDto,
  type ConnectInvestorGetAdvisorProfileResponseDto,
  ConnectInvestorRiskProfilingApi,
} from '@bambu/api-client';
import type { QueryClient, UseQueryOptions } from '@tanstack/react-query';
import { useQuery } from '@tanstack/react-query';

export const getInvestorRiskProfilesQuery = {
  queryKey: ['investorRiskProfiles'],
  queryFn: fetchInvestorRiskProfiles,
  staleTime: 1000 * 59 * 5,
};

export function useGetInvestorRiskProfiles<
  T = ConnectInvestorGetInvestorRiskProfilesResponseDto
>(
  options?: UseQueryOptions<
    ConnectInvestorGetInvestorRiskProfilesResponseDto,
    unknown,
    T,
    string[]
  >
) {
  return useQuery({
    queryKey: getInvestorRiskProfilesQuery.queryKey,
    queryFn: getInvestorRiskProfilesQuery.queryFn,
    staleTime: getInvestorRiskProfilesQuery.staleTime,
    ...options,
  });
}

export const getInvestorRiskProfilesLoader =
  (queryClient: QueryClient) =>
  async (): Promise<ConnectInvestorGetAdvisorProfileResponseDto> => {
    // return data or fetch it
    return (
      queryClient.getQueryData(getInvestorRiskProfilesQuery.queryKey) ??
      (await queryClient.fetchQuery(getInvestorRiskProfilesQuery))
    );
  };

async function fetchInvestorRiskProfiles() {
  const api = new ConnectInvestorRiskProfilingApi();
  const res = await api.getInvestorRiskProfiles();
  return res.data;
}

export default useGetInvestorRiskProfiles;
