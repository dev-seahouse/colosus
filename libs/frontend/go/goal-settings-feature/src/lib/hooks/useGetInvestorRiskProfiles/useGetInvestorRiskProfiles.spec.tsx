import {
  queryClientWrapper,
  renderHook,
  waitFor,
} from '@bambu/react-test-utils';
import useGetInvestorRiskProfiles from './useGetInvestorRiskProfiles';

describe('useGetInvestorRiskProfiles', () => {
  it('should render successfully', async () => {
    const { result } = renderHook(() => useGetInvestorRiskProfiles(), {
      wrapper: queryClientWrapper,
    });
    await waitFor(() => expect(result.current.isSuccess).toBe(true));
    expect(result.current.data).toMatchSnapshot();
  });
});
