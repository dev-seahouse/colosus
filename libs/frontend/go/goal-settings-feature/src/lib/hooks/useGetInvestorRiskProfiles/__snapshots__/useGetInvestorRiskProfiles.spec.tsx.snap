// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useGetInvestorRiskProfiles > should render successfully 1`] = `
[
  {
    "id": "8e6fb5fd-e8b4-4019-9d2c-6fd1ff8f6e5f",
    "lowerLimit": "1",
    "riskProfileDescription": "You don’t want to experience volatility in your portfolio.<br/>You don’t expect the value of your portfolio to go down at any significant period of time.<br/>You understand that expected returns are low",
    "riskProfileName": "Very Conservative",
    "tenantId": "0bc1e3e3-63fb-4495-beb8-411054baecda",
    "upperLimit": "1",
  },
  {
    "id": "eeef55b0-40a7-4980-b8ec-a94e01be1d13",
    "lowerLimit": "2",
    "riskProfileDescription": "You are OK with a bit of volatility in your portfolio.<br/>You understand that the value of your portfolio may go down for a short period of time before it bounces back.<br/>You understand that expected returns are below average.",
    "riskProfileName": "Conservative",
    "tenantId": "0bc1e3e3-63fb-4495-beb8-411054baecda",
    "upperLimit": "2",
  },
  {
    "id": "4bf6874c-e5ea-4f66-b2f9-e1212b823422",
    "lowerLimit": "3",
    "riskProfileDescription": "You are OK with some volatility in your portfolio.<br/>You understand that the value of your portfolio may go down for a moderate period of time before it bounces back.<br/>You understand that expected returns are average.",
    "riskProfileName": "Balanced",
    "tenantId": "0bc1e3e3-63fb-4495-beb8-411054baecda",
    "upperLimit": "3",
  },
  {
    "id": "e39f6a89-743f-4ed2-9307-22d5d545020c",
    "lowerLimit": "4",
    "riskProfileDescription": "You are OK with volatility in your portfolio.<br/>You understand that the value of your portfolio may go down for a significant period of time before it bounces back.<br/>You expect good returns in the mid to long term.",
    "riskProfileName": "Growth",
    "tenantId": "0bc1e3e3-63fb-4495-beb8-411054baecda",
    "upperLimit": "4",
  },
  {
    "id": "149a15e6-7152-4a2d-9d16-0d6a0c70b825",
    "lowerLimit": "5",
    "riskProfileDescription": "You are OK with high volatility in your portfolio.<br/>You understand that the value of your portfolio may go down sharply in the future but you know that you will reap big benefits if you are patient enough.<br/>You expect high returns in the long term.",
    "riskProfileName": "Aggressive",
    "tenantId": "0bc1e3e3-63fb-4495-beb8-411054baecda",
    "upperLimit": "5",
  },
]
`;
